### PR TITLE
feat: Redesign Container Queries to allow dynamic queries, and support the new Content Book

### PR DESCRIPTION
**Pull request by magicus**
I have some kind of proof of concept of populating our quest book on beta, but with lots of limitations. The content book must have the correct filter selected. Clicking on a quest to track/untrack is not supported. Additional requirements (profession) is not supported.

### CODE REVIEW

**Comment by kristofbolyai** [comment]
I don't fully get the concept of this PR. I thought we are building the dynamic query system first, but this PR is closer to shoehorning the new parsing into the old system. I would be glad to help out this PR, but first we need to have a chat what the next step is.

**Comment by magicus** [comment]
Your reading is absolutely correct. I started with this part to be sure that the dynamic rewrite will work, once I start writing it, since it is tricky to debug unless you get a proper result out of the query to look at.

This is not intended to be merged, it is just a show-case that this part of the parsing works.

File: common/src/main/java/com/wynntils/handlers/container/scriptedquery/QueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * 1) a startAction (to open the container)
+ * 2) a verification (to check the container that opened, if any)
+ * 3) a handleContent (to actually consume the content of the container)
+ *
+ * A QueryStep with no verification will assume no new container is opened.
+ * A QueryStep with no handleContent will just perform a no-op for consuming the content.
+ */
+public final class QueryBuilder {
+    private static final Consumer<String> DEFAULT_ERROR_HANDLER =
+            (errorMsg) -> WynntilsMod.warn("Error in ScriptedContainerQuery");
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/handlers/container/scriptedquery/QueryBuilder.java]:
```suggestion
            (errorMsg) -> WynntilsMod.warn("Error in ScriptedContainerQuery: " + errorMsg);
```

Why hide it?

File: common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java [hunk 2]
```diff
@@ -48,59 +50,68 @@ public void updateParsedAbilityTree() {
...
+                        QueryStep.clickOnSlot(NEXT_PAGE_SLOT).processIncomingContainer(processor::processPage))
+
+                // Go back to initial page
+                .repeat(
+                        c -> {
+                            // Go back as many pages as the original count was from the end
+                            this.pageCount++;
+                            return this.pageCount != Models.AbilityTree.ABILITY_TREE_PAGES;
+                        },
+                        QueryStep.clickOnSlot(PREVIOUS_PAGE_SLOT))
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java:94]: This "resetting" is new. Is this needed / does this slow queries down? I assume this is more important in the quest book filters...

**Review by kristofbolyai — COMMENTED** [review]
Just a quick status update. I started reviewing this, got through most of the system code. I started reviewing the new queries, but I am going to have to come back tomorrow to finish. 

Overall it is looking really great so far, can't wait to work with this :)

**Comment by kristofbolyai** [comment]
> Your reading is absolutely correct. I started with this part to be sure that the dynamic rewrite will work, once I start writing it, since it is tricky to debug unless you get a proper result out of the query to look at.
> 
> This is not intended to be merged, it is just a show-case that this part of the parsing works.

I would assume this comment to be out of date by now.

File: common/src/main/java/com/wynntils/handlers/container/scriptedquery/QueryBuilder.java (hunk 1 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/handlers/container/scriptedquery/QueryBuilder.java]: No reason. I was just stupid.

File: common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java (hunk 2 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/abilitytree/AbilityTreeContainerQueries.java:94]: Yes, it is new. I wouldn't say it slows down things much. In the worst case, you had the tree on the first page, then you need to go back 6 pages. My rough estimate is that 20 pages (the entire content book) takes 3-4 seconds, so at most an additional second.

I think we have to include your ability tree stuff for real to know if this becomes a problem. But I think it is proper of us to restore everything to the way it was, so our inspections are "hidden" for the user. So unless it is a serious problem I'd rather keep it. My thinking is that the reason this was not implemented before was that it was virtually impossible to do with the old design.

**Comment by magicus** [comment]
> > This is not intended to be merged, it is just a show-case that this part of the parsing works.
> 
> I would assume this comment to be out of date by now.

Yes, the branch meandered from a testing PoC to a fully fledged implementation, so I kept the same PR.

File: common/src/main/java/com/wynntils/models/content/ContentBookQueries.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                // Finally signal we're done
+                .execute(() -> processResult.accept(newContent, progress))
+                .build();
+
+        query.executeQuery();
+    }
+
+    private String getActiveFilter(ItemStack itemStack) {
+        StyledText itemName = InventoryUtils.getItemName(itemStack);
+        if (!itemName.equals(StyledText.fromString("§eFilter"))) return null;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/content/ContentBookQueries.java:133]: Extract to a const?

File: common/src/main/java/com/wynntils/models/content/ContentBookQueries.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+            ContentInfo contentInfo = contentItemOpt.get().getContentInfo();
+
+            newContent.add(contentInfo);
+        }
+    }
+
+    protected void toggleTracking(String name, ContentType contentType) {
+        // We do not want to change filtering when tracking, since we get
+        // no chance to reset it
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/content/ContentBookQueries.java:160]: I am not sure why this is the case?

File: common/src/main/java/com/wynntils/models/quests/QuestModel.java [hunk 5]
```diff
@@ -106,11 +112,12 @@ public List<List<StyledText>> getDialogueHistory() {
...
 
     public void startTracking(QuestInfo questInfo) {
-        CONTAINER_QUERIES.toggleTracking(questInfo);
+        Models.Content.startTracking(
+                questInfo.getName(), questInfo.isMiniQuest() ? ContentType.MINI_QUEST : ContentType.QUEST);
     }
 
     public void stopTracking() {
-        McUtils.sendCommand("tracking");
+        Models.Content.stopTracking();
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/quests/QuestModel.java:120]: Did this break? I would assume the command to be faster...

File: common/src/main/java/com/wynntils/models/quests/QuestModel.java [hunk 6]
```diff
@@ -143,26 +148,56 @@ private String stripPrefix(String name) {
...
+    private static QuestInfo getQuestInfoFromContent(ContentInfo content) {
+        // We should always have a length, but if not, better fake one than crashing
+
+        return new QuestInfo(
+                content.name(),
+                QuestStatus.fromContentStatus(content.status()),
+                QuestLength.fromContentLength(content.length()),
+                content.requirements().level().key(),
+                content.description().orElse(StyledText.EMPTY),
+                // FIXME! Additional requirements missing
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/quests/QuestModel.java:191]: Speaks for itself

**Review by kristofbolyai — APPROVED** [review]
Allow of my comments here are something I want to address later. Right now the focus is getting this out.

File: common/src/main/java/com/wynntils/models/content/ContentBookQueries.java (hunk 4 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/content/ContentBookQueries.java:160]: As soon as you click an item to track it, the container closes, but any changes we have made to the filter state will remain.

File: common/src/main/java/com/wynntils/models/quests/QuestModel.java (hunk 5 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/quests/QuestModel.java:120]: Yes, the command stopped working. :-(

File: common/src/main/java/com/wynntils/models/quests/QuestModel.java (hunk 6 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/quests/QuestModel.java:191]: Yeah, I forgot about that. But even if I hadn't, I wouldn't have had the time to fix it. Tracked in https://github.com/Wynntils/Artemis/issues/1730.

File: common/src/main/java/com/wynntils/models/content/ContentBookQueries.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/content/ContentBookQueries.java:160]: This I failed to observe, and added reset code to tracking... (although it is disabled atm)

File: common/src/main/java/com/wynntils/models/content/ContentBookQueries.java (hunk 3 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/content/ContentBookQueries.java:133]: Fixed in #1744.

**Pull request merged**